File: go/internal/fndsa/params.go
package fndsa

// Q is the modulus used in FN-DSA (FIPS 206): q = 12289.
const Q = 12289

// Params holds the parameter set for an FN-DSA security level.
type Params struct {
	Name   string
	N      int     // polynomial degree (512 or 1024)
	LogN   int     // log2(N)
	Sigma  float64 // Gaussian standard deviation
	PKSize int     // public key size in bytes
	SKSize int     // secret key size in bytes
	// SigSize is the authoritative signature size used by callers:
	//   - for non-padded variants: maximum variable-length signature (SigMaxLen)
	//   - for padded variants: fixed padded-signature length
	SigSize int
	// SigMaxLen is the max variable-length signature before padding. Equals SigSize for non-padded variants.
	SigMaxLen int
	Padded    bool // true for FN-DSA-PADDED variants
	BetaSq    int  // β² bound for signature validity
}

// Pre-defined parameter sets per FIPS 206 Table 2.
var (
	// FNDSA512 is FN-DSA-512 (FALCON-512), targeting NIST security level 1.
	FNDSA512 = &Params{
		Name:      "FN-DSA-512",
		N:         512,
		LogN:      9,
		Sigma:     165.736617183,
		PKSize:    897,
		SKSize:    1281,
		SigSize:   666,
		SigMaxLen: 666,
		Padded:    false,
		BetaSq:    34034726,
	}

	// FNDSA1024 is FN-DSA-1024 (FALCON-1024), targeting NIST security level 5.
	FNDSA1024 = &Params{
		Name:      "FN-DSA-1024",
		N:         1024,
		LogN:      10,
		Sigma:     168.388571447,
		PKSize:    1793,
		SKSize:    2305,
		SigSize:   1280,
		SigMaxLen: 1280,
		Padded:    false,
		BetaSq:    70265242,
	}

	// FNDSAPadded512 is FN-DSA-PADDED-512 (FALCON-PADDED-512): same security
	// level as FNDSA512 but with a fixed-length padded signature format.
	FNDSAPadded512 = &Params{
		Name:      "FN-DSA-PADDED-512",
		N:         512,
		LogN:      9,
		Sigma:     165.736617183,
		PKSize:    897,
		SKSize:    1281,
		SigSize:   809,
		SigMaxLen: 666,
		Padded:    true,
		BetaSq:    34034726,
	}

	// FNDSAPadded1024 is FN-DSA-PADDED-1024 (FALCON-PADDED-1024): same
	// security level as FNDSA1024 but with a fixed-length padded signature.
	FNDSAPadded1024 = &Params{
		Name:      "FN-DSA-PADDED-1024",
		N:         1024,
		LogN:      10,
		Sigma:     168.388571447,
		PKSize:    1793,
		SKSize:    2305,
		SigSize:   1473,
		SigMaxLen: 1280,
		Padded:    true,
		BetaSq:    70265242,
	}
)

// AllParams returns all defined FN-DSA parameter sets.
func AllParams() []*Params {
	return []*Params{FNDSA512, FNDSA1024, FNDSAPadded512, FNDSAPadded1024}
}

File: swift/fndsa/Sources/FnDsa/Params.swift
// Params.swift — FN-DSA parameter sets (FIPS 206 / FALCON)

public let Q: Int32 = 12289

public struct Params {
    public let name: String
    public let n: Int
    public let logN: Int
    public let padded: Bool
    public let pkSize: Int
    public let skSize: Int
    public let sigSize: Int
    /// Maximum variable-length signature size (before padding).
    /// For non-padded variants this equals sigSize; for padded variants it is
    /// the non-padded maximum.
    public let sigMaxLen: Int
    public let betaSq: Int64
    public let fgBits: Int
}

public let fnDsa512 = Params(
    name: "FN-DSA-512", n: 512, logN: 9, padded: false,
    pkSize: 897, skSize: 1281, sigSize: 666, sigMaxLen: 666,
    betaSq: 34034726, fgBits: 6)

public let fnDsa1024 = Params(
    name: "FN-DSA-1024", n: 1024, logN: 10, padded: false,
    pkSize: 1793, skSize: 2305, sigSize: 1280, sigMaxLen: 1280,
    betaSq: 70265242, fgBits: 5)

public let fnDsaPadded512 = Params(
    name: "FN-DSA-PADDED-512", n: 512, logN: 9, padded: true,
    pkSize: 897, skSize: 1281, sigSize: 809, sigMaxLen: 666,
    betaSq: 34034726, fgBits: 6)

public let fnDsaPadded1024 = Params(
    name: "FN-DSA-PADDED-1024", n: 1024, logN: 10, padded: true,
    pkSize: 1793, skSize: 2305, sigSize: 1473, sigMaxLen: 1280,
    betaSq: 70265242, fgBits: 5)
